src/app/main.rs

```diff
@@ -0,0 +1,545 @@
+            config.inbound_ports_disable_protocol_detection,
+        );
+        info!(
+            "protocol detection disabled for outbound ports {:?}",
+            config.outbound_ports_disable_protocol_detection,
+        );
+
+        let tap_next_id = tap::NextId::default();
+        let (taps, observe) = control::Observe::new(100);
+        let (http_metrics, http_report) = proxy::http::metrics::new::<
+            EndpointLabels,
+            classify::Class,
+        >(config.metrics_retain_idle);
+        let (transport_metrics, transport_report) = transport::metrics::new();
+
+        let (tls_config_sensor, tls_config_report) = telemetry::tls_config_reload::new();
+
+        let report = telemetry::Report::new(
+            http_report,
+            transport_report,
+            tls_config_report,
+            telemetry::process::Report::new(start_time),
+        );
+
+        let tls_client_config = tls_config_watch.client.clone();
+        let tls_cfg_bg = tls_config_watch.start(tls_config_sensor);
+
+        let controller_tls = config.tls_settings.as_ref().and_then(|settings| {
+            settings
+                .controller_identity
+                .as_ref()
+                .map(|controller_identity| tls::ConnectionConfig {
+                    server_identity: controller_identity.clone(),
+                    config: tls_client_config.clone(),
+                })
+        });
+
+        let (dns_resolver, dns_bg) = dns::Resolver::from_system_config_and_env(&config)
+            .unwrap_or_else(|e| {
+                // TODO: DNS configuration should be infallible.
+                panic!("invalid DNS configuration: {:?}", e);
+            });
+
+        let (resolver, resolver_bg) = control::destination::new(
+            dns_resolver.clone(),
+            config.namespaces.clone(),
+            control_host_and_port,
+            controller_tls,
+            config.control_backoff_delay,
+            config.destination_concurrency_limit,
+        );
+
+        const MAX_IN_FLIGHT: usize = 10_000;
+
+        let (drain_tx, drain_rx) = drain::channel();
+
+        let outbound = {
+            use super::outbound;
+
+            let http_metrics = http_metrics.clone();
+
+            // As the outbound proxy accepts connections, we don't do any
+            // special transport-level handling.
+            let accept = transport_metrics.accept("outbound").bind(());
+
+            // Establishes connections to remote peers.
+            let connect = transport_metrics
+                .connect("outbound")
+                .and_then(proxy::timeout::Layer::new(config.outbound_connect_timeout))
+                .bind(connect::Stack::new());
+
+            // As HTTP requests are accepted, we add some request extensions
+            // including metadata about the request's origin.
+            let source_layer =
+                timestamp_request_open::Layer::new().and_then(insert_target::Layer::new());
```

Comment:
Yeah, i think the free function approach is probably the best option.

Reply:
Yeah, I took a quick pass at at blocking out what a trait that bundles together a stack and layer type would look like, and I think it may be more trouble than it's worth.